discovery/event.go

```diff
@@ -0,0 +1,39 @@
+package discovery
+
+import "sync"
+
+// event is a one time event that can be marked done.
+type event struct {
+	done chan struct{}
+	once sync.Once
+}
+
+func newEvent() *event {
+	return &event{
+		done: make(chan struct{}),
+	}
+}
+
+// SetDone marks the event as done.
+// After this, Done() returns a closed channel.
+// This is safe to call multiple times, concurrently.
+func (e *event) SetDone() {
+	e.once.Do(func() {
+		close(e.done)
+	})
+}
+
+// Done returns a channel that is closed once this event is done.
+func (e *event) Done() <-chan struct{} {
+	return e.done
+}
+
+// IsDone returns true if the event is done.
+func (e *event) IsDone() bool {
+	select {
+	case <-e.Done():
+		return true
+	default:
+		return false
+	}
+}
```

Comment:
I like this type a lot! We could also add a `Wait` method, similar to memdb's `WatchSet`:

```Go
func (e *event) Wait(ctx context.Context) error {
	select {
	case <-e.Done():
		return nil
	case <-ctx.Done():
		return ctx.Err()
	}
}
```

Totally optional, I think it's clear as it is.